app/uk/gov/hmrc/mtdsatestsupportapi/controllers/requestParsers/validators/CreateTestBusinessValidator.scala

```diff
@@ -90,32 +90,88 @@ class CreateTestBusinessValidator @Inject() (clock: Clock, featureSwitches: Feat
+        Seq(MissingPostcodeError)
+      } else {
+        Nil
+      }
+    } else {
+      (business.businessAddressPostcode, business.businessAddressCountryCode) match {
+        case (None, Some("GB")) => List(MissingPostcodeError)
+        case _                  => NoValidationErrors
+      }
     }
   }
 
-  private def validateAccountingPeriod(business: Business): Seq[MtdError] =
-    if (featureSwitches.isEnabled("release5"))
+  private def validateAccountingPeriod(business: Business): Seq[MtdError] = {
+    if (featureSwitches.isEnabled("release5")) {
       (business.firstAccountingPeriodStartDate, business.firstAccountingPeriodEndDate) match {
         case (Some(start), Some(end)) => TaxYearAlignmentDateRangeValidation.validate(start, end, RuleFirstAccountingDateRangeInvalid)
         case (None, Some(_))          => Seq(MissingFirstAccountintPeriodStartDateError)
         case (Some(_), None)          => Seq(MissingFirstAccountintPeriodEndDateError)
         case _                        => Nil
       }
-    else Nil
+    } else {
+      Nil
+    }
+  }
 
   private def validateCommencementDate(date: LocalDate): List[MtdError] = {
     val today = LocalDate.now(clock)
 
     if (date < today) Nil else List(RuleCommencementDateNotSupported)
   }
 
+  private def validateTradingName(business: Business): Seq[MtdError] = {
+    if (featureSwitches.isEnabled("release5")) {
+      if (forbidsTradingName(business) && business.tradingName.nonEmpty) {
+        Seq(RuleUnexpectedTradingName)
+      } else if (requiresTradingName(business) && business.tradingName.isEmpty) {
+        Seq(RuleMissingTradingName)
+      } else {
+        Nil
+      }
+    } else {
+      Nil
+    }
+  }
+
+  private def validateBusinessAddress(business: Business): Seq[MtdError] = {
+    def haveSufficientBusinessAddress = business.businessAddressLineOne.nonEmpty && business.businessAddressCountryCode.nonEmpty
+
+    if (featureSwitches.isEnabled("release5")) {
+      if (forbidsBusinessAddress(business) && business.hasAnyBusinessAddressDetails) {
+        Seq(RuleUnexpectedBusinessAddress)
+      } else if (requiresBusinessAddress(business) && !haveSufficientBusinessAddress) {
+        Seq(RuleMissingBusinessAddress)
+      } else {
+        Nil
+      }
+    } else {
+      Nil
+    }
+  }
+
+  private def requiresBusinessPostcode(business: Business): Boolean = {
+    !forbidsBusinessAddress(business) &&
+      business.businessAddressCountryCode.contains("GB")
+  }
+
+  private def requiresBusinessAddress(business: Business): Boolean = business.typeOfBusiness.isSelfEmployment
+
+  private def forbidsBusinessAddress(business: Business): Boolean = business.typeOfBusiness.isProperty
+
+  private def requiresTradingName(business: Business): Boolean = business.typeOfBusiness.isSelfEmployment
+
+  private def forbidsTradingName(business: Business): Boolean = business.typeOfBusiness.isProperty
```

Comment:
Since there are effectively 4 methods doing 2 things, and the fields are clearly named, could we just inline the field selections? e.g. if `if (forbidsBusinessAddress(business) && business.hasAnyBusinessAddressDetails)` becomes `if (business.typeOfBusiness.isProperty && business.hasAnyBusinessAddressDetails)`, I think it is quite easy to deduce that business address details are forbidden for a property business.

Reply:
I saw the methods as a bit of self-documentation about the logic behind the trading names and business addresses. They're 4 things in the docs, so I wanted 4 descriptions in the code.

But I'm happy to inline all 4 of them if you prefer.

I don't see them as "2 things", even though the implementation happens to be the same for now. Requiring trade names is to me unrelated to requiring business addresses. So I wouldn't want to combine the apparent duplicates.

It also made the logic more readable to me. Easier to spot bugs when every piece of logic matched the docs. I appreciate that that's subjective.

Want me to inline all 4 of them?

Reply:
Maybe this is comes down to a preference thing again (?). Happy for you to leave if so.

Reply:
I tried to inline them just now, but found that `forbidsBusinessAddress` was used in two places that I wanted to to share code. (normal address validation and postcode validation need to agree on when a business address is forbidden)

Since I can't remove all of them, I'll keep all 4 of them for consistency.